[Feature] Integrate test result cells to home screen

johannesrohwer opened the pull request:
<!-- 
Thank you for supporting us with your Pull Request! 🙌 ❤️  
Before submitting, please take the time to check the points below and provide some descriptive information.
-->

## Description
This PR implements the result card feature that allows to see status information about tests on the home screen. Furthermore, I do some refactoring of the `HomeViewController` and the `HomeInteractor`. 

## Checklist

* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [ ] If this PR comes from a fork, please [Allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
* [x] Set a speaking title. Format: {task_name} (closes #{issue_number}). For example: Use logger (closes # 41)
* [ ] [Link your Pull Request to an issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) (if applicable)
* [ ] Create Work In Progress [WIP] pull requests only if you need clarification or an explicit review before you can continue your work item.
* [x] Make sure that your PR is not introducing _unnecessary_ reformatting (e.g., introduced by on-save hooks in your IDE)

## Description
<!-- Please be brief in describing which issue is solved by your PR or which enhancement it brings -->

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Views/Home Screen/Cells/HomeTestResultCell.swift:
```diff
@@ -31,99 +31,9 @@ class HomeTestResultCell: HomeCardCollectionViewCell {
-
-	weak var delegate: TestResultCollectionViewCellDelegate?
+	weak var delegate: HomeTestResultCellDelegate?
 
```
IB is crashed when it tries to open .xib 

![image](https://user-images.githubusercontent.com/61236601/83555371-79371680-a50e-11ea-9c54-97ca7e7944c2.png)

johannesrohwer replied:
had the same issue from time to time, i thought it was only my machine but i will see if i can find a reason why xcode does not like the xib

aleksandr-tikhonov replied:
I think, the reason is in some ibdesignable view in this .xib

johannesrohwer replied:
i could not figure out what causes the issue. i can open and edit it normally and sometimes it just crashes. :(

ChristianKienle replied:
No crash on my end… not sure what to do. The change @johannesrohwer did was not that big and does not look suspicious.

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Views/Home Screen/Cells/HomeTestResultCell.swift:
```diff
@@ -31,99 +31,9 @@ class HomeTestResultCell: HomeCardCollectionViewCell {
-	weak var delegate: TestResultCollectionViewCellDelegate?
+	weak var delegate: HomeTestResultCellDelegate?
 
 	@IBAction func buttonTapped(_: UIButton) {
```
I think, it's the same cell as `SubmitCollectionViewCell`, so please remove duplication.

johannesrohwer replied:
I added a `HomeCardCellButtonDelegate` that replaced the duplicated delegates

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionNavigationController.swift:
```diff
@@ -185,6 +185,7 @@ class ExposureSubmissionNavigationController: UINavigationController, UINavigati
 
 	override func viewWillDisappear(_ animated: Bool) {
 		super.viewWillAppear(animated)
+		self.homeViewController?.updateTestResultState()
```
Can we do it in another way?

johannesrohwer replied:
is there a way that you would prefer? this was just the simplest way to ensure the home view is updated whenever we close the modally presented exposure submission navigator.

aleksandr-tikhonov replied:
you can review for example `stateHandler` in interactor and do it in similar way

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Scenes/Home/HomeInteractor.swift:
```diff
@@ -18,9 +18,12 @@
-// swiftlint:disable:next type_body_length
 final class HomeInteractor {
 
+	typealias SectionDefinition = (section: HomeViewController.Section, cellConfigurators: [CollectionViewCellConfiguratorAny])
```
👍

ChristianKienle replied:
🥳

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Scenes/Home/HomeViewController.swift:
```diff
@@ -335,6 +339,7 @@ final class HomeViewController: UIViewController {
 		collectionView.translatesAutoresizingMaskIntoConstraints = false
 		collectionView.isAccessibilityElement = false
 		collectionView.shouldGroupAccessibilityChildren = true
+		collectionView.alwaysBounceVertical = true
```
hm, why?

johannesrohwer replied:
feels nicer in my opinion, otherwise the app feels very static when there are few cells on the screen.. but this is just my personal preference :)

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Scenes/Home/HomeViewController.swift:
```diff
@@ -403,21 +415,35 @@ final class HomeViewController: UIViewController {
+
+	private func fetchTestResult() {
+		DispatchQueue.global(qos: .userInteractive).async {
+			self.fetchTestResultHelper()
+		}
+	}
+
+	private func fetchTestResultHelper() {
+		// Make sure we are able to fetch.
+		guard store.registrationToken != nil else { return }
 		self.exposureSubmissionService?.getTestResult { result in
 			switch result {
 			case .failure(let error):
-				appLogger.log(message: "Could not update test state: \(error)", file: #file, line: #line, function: #function)
+				appLogger.log(message: "Error while fetching result: \(error)", file: #file, line: #line, function: #function)
 			case .success(let result):
 				self.testResult = result
-				configurator.configure(cell: cell)
+				self.homeInteractor.reloadTestResult(with: result)
 			}
 		}
 	}
```
shouldn't be here

johannesrohwer replied:
this was moved to the home interactor

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Scenes/Home/HomeViewController.swift:
```diff
@@ -403,21 +415,35 @@ final class HomeViewController: UIViewController {
+		fetchTestResult()
+	}
+
+	private func fetchTestResult() {
```
I lost one of my comment

aleksandr-tikhonov commented on src/xcode/ENA/ENA/Source/Scenes/Home/HomeViewController.swift:
```diff
@@ -403,21 +415,35 @@ final class HomeViewController: UIViewController {
+
+	private func fetchTestResult() {
+		DispatchQueue.global(qos: .userInteractive).async {
+			self.fetchTestResultHelper()
+		}
+	}
+
+	private func fetchTestResultHelper() {
+		// Make sure we are able to fetch.
+		guard store.registrationToken != nil else { return }
 		self.exposureSubmissionService?.getTestResult { result in
 			switch result {
 			case .failure(let error):
-				appLogger.log(message: "Could not update test state: \(error)", file: #file, line: #line, function: #function)
+				appLogger.log(message: "Error while fetching result: \(error)", file: #file, line: #line, function: #function)
 			case .success(let result):
 				self.testResult = result
-				configurator.configure(cell: cell)
+				self.homeInteractor.reloadTestResult(with: result)
 			}
 		}
 	}
```
business logic should be in interactor or in another class, not in view controller

johannesrohwer replied:
i moved that to the interactor

aleksandr-tikhonov commented:
Could you please write a short comment in my comments for example "I fixed it", that I could see what you fixed and what not :)

johannesrohwer commented:
> Could you please write a short comment in my comments for example "I fixed it", that I could see what you fixed and what not :)

I added some comments to clear up what I changed. I am now waiting for your PR to be merged to master and will integrate your changes into this branch. When I am done, I will mark this PR ready for review and request another review round. Thanks for your feedback so far!
